### ISSUE TITLE
Ejabberd Clustor file sharing not working for one node to second node

### ISSUE THREAD

**Issue by kamalakkannan1984** [processone/ejabberd#3359]
## Environment

- ejabberd version: 20.04
- Erlang version: `erl +V`
- OS: Linux (Debian)
- Installed from: source | distro package | official deb/rpm | official binary installer | other

## Configuration (only if needed): grep -Ev '^$|^\s*#' ejabberd.yml

```yaml
loglevel: 4
...
```

## Errors from error.log/crash.log
2020-08-19 10:57:54.010 [warning] <0.1997.0>@mod_http_upload:process:563 Rejecting file ca427369-3812-4459-8162-ba4694f831fd.jpg from ::ffff:127.0.0.1 for 7049.xxxx.yyyyy.com: Invalid slot

Ejabberd Clustor file sharing not working for one node to second node

## Bug description

The Ejabberd cluster setup via HAProxy load balancer. In the cluster setup I have Node1(Ejabberd server 1) and Node2(Ejabberd server 2). When I upload file some time get rejected regarding invalid slot.

Our setup is HAProxy connect with two nodes(Node1 and Node2) 

Note: Single Node(Ejabberd server) working fine.

Please any one share your ideas and feedback.

**Comment by prefiks** [comment]
Yes indeed there is no sharing token assignments between nodes in this module, so you could be only be able to perform upload on node where slot allocation was performed. To make this work we would need to move slot keeping logic to outside each module state, possibly in memory mnesia db?

**Comment by weiss** [comment]
> To make this work we would need to move slot keeping logic to outside each module state, possibly in memory mnesia db?

My plan is to get rid of the module state altogether by replacing it with the [`external_secret`][1] mechanism.

[1]: https://modules.prosody.im/mod_http_upload_external

**Comment by weiss** [comment]
@kamalakkannan1984 right now my suggestion would be to have an external web server handle the actual file uploads by using one of the [implementations mentioned here][1].  If you specify the same [external_secret][2] for `mod_http_upload` and the web servers running on your nodes, uploads will be accepted by all of them.

[1]: https://modules.prosody.im/mod_http_upload_external#implementations
[2]: https://docs.ejabberd.im/admin/configuration/modules/#mod-http-upload

**Comment by mkl262** [comment]
The same thing happens to me on ejabberd 19.09.1.
I have a similar setup with 2 nodes and a HAProxy server.

**Comment by weiss** [comment]
Sure.  The [suggested solution][1] would work with 19.09.1 as well.

[1]: https://github.com/processone/ejabberd/issues/3359#issuecomment-677506724

**Comment by catap** [comment]
> @kamalakkannan1984 right now my suggestion would be to have an external web server handle the actual file uploads by using one of the [implementations mentioned here](https://modules.prosody.im/mod_http_upload_external#implementations). If you specify the same [external_secret](https://docs.ejabberd.im/admin/configuration/modules/#mod-http-upload) for `mod_http_upload` and the web servers running on your nodes, uploads will be accepted by all of them.

Here two catches as I see:
 1. You should ditribute files across the cluster by hand.
 2. You can't use global hostname of upload node, instead you should vhost per each vhost in ejabberd.
 
 Am I wrong?
